File: third-party/inputtino/bindings/python/src/inputtino/mouse.py
"""Mouse input device implementation."""

import time

from . import _core
from ._core import MouseButton
from .base import DeviceDefinition, VirtualDevice

# Default device definition for mouse
DEFAULT_MOUSE = DeviceDefinition(
    name="Wolf mouse virtual device",
    vendor_id=0xAB00,
    product_id=0xAB01,
    version=0xAB00,
)


class Mouse(VirtualDevice):
    """Virtual mouse input device.

    This class provides functionality to simulate mouse movements,
    clicks, and scrolling. The device appears as a real mouse input
    device to the system.
    """

    def __init__(self, device_def: DeviceDefinition = DEFAULT_MOUSE) -> None:
        """Initialize the mouse device with optional custom device definition.

        Args:
            device_def: Optional device definition for customizing the virtual device properties

        Raises:
            RuntimeError: If device creation fails
        """
        self._mouse = _core.Mouse.create(device_def.to_core())
        super().__init__(self._mouse)

    def move(self, delta_x: int, delta_y: int) -> None:
        """Move the mouse cursor relative to its current position.

        Args:
            delta_x: Horizontal movement (positive is right, negative is left)
            delta_y: Vertical movement (positive is down, negative is up)
        """
        self._mouse.move(delta_x, delta_y)

    def move_abs(self, x: int, y: int, screen_width: int, screen_height: int) -> None:
        """Move the mouse cursor to an absolute screen position.

        Args:
            x: Target X coordinate
            y: Target Y coordinate
            screen_width: Width of the screen in pixels
            screen_height: Height of the screen in pixels
        """
        self._mouse.move_abs(x, y, screen_width, screen_height)

    def press(self, button: MouseButton) -> None:
        """Press a mouse button.

        Args:
            button: The button to press
        """
        self._mouse.press(button)

    def release(self, button: MouseButton) -> None:
        """Release a mouse button.

        Args:
            button: The button to release
        """
        self._mouse.release(button)

    def scroll_vertical(self, distance: int) -> None:
        """Scroll vertically.

        Args:
            distance: Scroll distance in high resolution units (±120 per notch).
                     Positive values scroll down, negative values scroll up.
        """
        self._mouse.vertical_scroll(distance)

    def scroll_horizontal(self, distance: int) -> None:
        """Scroll horizontally.

        Args:
            distance: Scroll distance in high resolution units (±120 per notch).
                     Positive values scroll right, negative values scroll left.
        """
        self._mouse.horizontal_scroll(distance)

    def click(
        self, button: MouseButton = MouseButton.LEFT, duration: float = 0.0
    ) -> None:
        """Perform a mouse click by pressing and releasing a button.

        Args:
            button: The button to click
            duration: How long to hold the button down in seconds
        """
        try:
            self.press(button)
            if duration > 0:
                time.sleep(duration)
        finally:
            self.release(button)

File: third-party/inputtino/bindings/rust/inputtino/src/mouse.rs
use std::path::PathBuf;

use crate::common::{get_nodes, make_device, DeviceDefinition};
use crate::sys::{
    inputtino_mouse_create, inputtino_mouse_destroy, inputtino_mouse_get_nodes,
    inputtino_mouse_move, inputtino_mouse_move_absolute, inputtino_mouse_press_button,
    inputtino_mouse_release_button, inputtino_mouse_scroll_horizontal,
    inputtino_mouse_scroll_vertical,
};

use crate::{InputtinoError, MouseButton};

pub struct Mouse {
    mouse: *mut crate::sys::InputtinoMouse,
}

impl Mouse {
    /// Create a new emulated mouse device with the given device definition.
    ///
    /// # Examples
    ///
    /// ```
    /// let definition = inputtino::DeviceDefinition::new(
    ///     "Inputtino Mouse",
    ///     0x0,
    ///     0x0,
    ///     0x0,
    ///     "00:11:22:33:44",
    ///     "00:11:22:33:44",
    /// );
    /// let device = inputtino::Mouse::new(&definition);
    /// ```
    pub fn new(device: &DeviceDefinition) -> Result<Self, InputtinoError> {
        make_device(inputtino_mouse_create, device).map(|mouse| Mouse { mouse })
    }

    pub fn get_nodes(&self) -> Result<Vec<PathBuf>, InputtinoError> {
        get_nodes(inputtino_mouse_get_nodes, self.mouse)
    }

    /// Move the mouse relative to its current position.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.move_rel(10, -10);
    /// ```
    pub fn move_rel(&self, x: i32, y: i32) {
        unsafe {
            inputtino_mouse_move(self.mouse, x, y);
        }
    }

    /// Move the mouse to absolute coordinates on the screen.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.move_abs(100, 100, 1920, 1080);
    /// ```
    pub fn move_abs(&self, x: i32, y: i32, screen_width: i32, screen_height: i32) {
        unsafe {
            inputtino_mouse_move_absolute(self.mouse, x, y, screen_width, screen_height);
        }
    }

    /// Simulate a button press.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.press_button(inputtino::MouseButton::LEFT);
    /// ```
    pub fn press_button(&self, button: MouseButton) {
        unsafe {
            inputtino_mouse_press_button(self.mouse, button);
        }
    }

    /// Simulate the release of a button.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.release_button(inputtino::MouseButton::LEFT);
    /// ```
    pub fn release_button(&self, button: MouseButton) {
        unsafe {
            inputtino_mouse_release_button(self.mouse, button);
        }
    }

    /// Simulate scrolling vertically for an amount of ticks.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.scroll_vertical(10);
    /// ```
    pub fn scroll_vertical(&self, amount: i32) {
        unsafe {
            inputtino_mouse_scroll_vertical(self.mouse, amount);
        }
    }

    /// Simulate scrolling horizontally for an amount of ticks.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// device.scroll_horizontal(-10);
    /// ```
    pub fn scroll_horizontal(&self, amount: i32) {
        unsafe {
            inputtino_mouse_scroll_horizontal(self.mouse, amount);
        }
    }
}

impl Drop for Mouse {
    fn drop(&mut self) {
        unsafe {
            inputtino_mouse_destroy(self.mouse);
        }
    }
}

unsafe impl Send for Mouse {}

#[cfg(test)]
mod tests {
    use crate::common::error_handler_fn;
    use std::ffi::CString;

    use super::*;

    #[test]
    fn test_inputtino_c_mouse() {
        let device_name = CString::new("Rusty Mouse").unwrap();
        let device_phys = CString::new("Rusty Mouse Phys").unwrap();
        let device_uniq = CString::new("Rusty Mouse Uniq").unwrap();
        let def = crate::sys::InputtinoDeviceDefinition {
            name: device_name.as_ptr(),
            vendor_id: 0,
            product_id: 0,
            version: 0,
            device_phys: device_phys.as_ptr(),
            device_uniq: device_uniq.as_ptr(),
        };
        // TODO: test this somehow
        let error_str = std::ptr::null_mut();
        let error_handler = crate::sys::InputtinoErrorHandler {
            eh: Some(error_handler_fn),
            user_data: error_str,
        };

        unsafe {
            let mouse = inputtino_mouse_create(&def, &error_handler);
            assert!(!mouse.is_null());

            let mut nodes_count: core::ffi::c_int = 0;
            let nodes = inputtino_mouse_get_nodes(mouse, &mut nodes_count);
            assert_eq!(nodes_count, 2);
            assert!(!nodes.is_null());
            // Check that the nodes start with /dev/input/event
            assert!(CString::from_raw(*nodes.offset(0))
                .to_str()
                .unwrap()
                .starts_with("/dev/input/event"));
            assert!(CString::from_raw(*nodes.offset(1))
                .to_str()
                .unwrap()
                .starts_with("/dev/input/event"));

            inputtino_mouse_destroy(mouse);
        }
    }
}
